crates/bevy_math/src/curves/variable_linear.rs

```diff
@@ -0,0 +1,210 @@
+use super::Curve;
+use crate::interpolation::Lerp;
+
+// TODO: Curve/Clip need a validation during deserialization because they are
+// structured as SOA (struct of arrays), so the vec's length must match
+
+// TODO: impl Serialize, Deserialize
+/// Curve with sparse keyframes frames, in another words a curve with variable frame rate;
+///
+/// Very useful curve, because can accommodate the output of a linear reduction keyframe algorithm
+/// to lower the memory foot print. But as a down side require the use of keyframe cursor, and
+/// loses performance when the curve frame rate is higher than the curve sampling frame rate
+#[derive(Default, Debug)]
+pub struct CurveVariableLinear<T> {
+    time_stamps: Vec<f32>,
+    keyframes: Vec<T>,
+}
+
+impl<T: Clone> Clone for CurveVariableLinear<T> {
+    fn clone(&self) -> Self {
+        Self {
+            time_stamps: self.time_stamps.clone(),
+            keyframes: self.keyframes.clone(),
+        }
+    }
+}
+
+impl<T> CurveVariableLinear<T> {
+    pub fn new(samples: Vec<f32>, values: Vec<T>) -> Self {
+        // TODO: Result?
+
+        // Make sure both have the same length
+        assert!(
+            samples.len() == values.len(),
+            "samples and values must have the same length"
+        );
+
+        assert!(
+            values.len() <= u16::MAX as usize,
+            "limit of {} keyframes exceeded",
+            u16::MAX
+        );
+
+        assert!(!samples.is_empty(), "empty curve");
+
+        // Make sure the
+        assert!(
+            samples
+                .iter()
+                .zip(samples.iter().skip(1))
+                .all(|(a, b)| a < b),
+            "time samples must be on ascending order"
+        );
+
+        Self {
+            time_stamps: samples,
+            keyframes: values,
+        }
+    }
+
+    pub fn from_line(t0: f32, t1: f32, v0: T, v1: T) -> Self {
+        Self {
+            time_stamps: if t1 >= t0 { vec![t0, t1] } else { vec![t1, t0] },
+            keyframes: vec![v0, v1],
+        }
+    }
+
+    pub fn from_constant(v: T) -> Self {
+        Self {
+            time_stamps: vec![0.0],
+            keyframes: vec![v],
+        }
+    }
+
+    // pub fn insert(&mut self, time_sample: f32, value: T) {
+    // }
+
+    // pub fn remove(&mut self, index: usize) {
+    //assert!(samples.len() > 1, "curve can't be empty");
+    // }
```

Comment:
and this?

Reply:
I'm rounding up the public api, next commit they will be implemented